### PR TITLE
Feat/label merge v2

### PR DESCRIPTION
**Pull request by jh3y**
- Different approach to #13 
- Implements custom React hook so that users can use a style and label value provided from a hook
- Requires storybook upgrade etc.

### CODE REVIEW

File: examples/LabeledMerge.tsx [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+        top: '-28px',
+        color: '#fff',
+        fontWeight: 'bold',
+        fontSize: '14px',
+        fontFamily: 'Arial,Helvetica Neue,Helvetica,sans-serif',
+        padding: '4px',
+        borderRadius: '4px',
+        backgroundColor: '#548BF4',
+        whiteSpace: 'nowrap',
+        ...(style as {})
```
Comment (tajo) [inline examples/LabeledMerge.tsx]:
```suggestion
        ...(style as React.CSSProperties)
```

File: examples/LabeledMerge.tsx [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+    </div>
+  );
+};
+class Labeled extends React.Component {
+  state: {
+    values: number[];
+    overlaps: any[];
+  } = {
+    values: [25, 50, 70, 80, 90],
+    overlaps: []
```
Comment (tajo) [inline examples/LabeledMerge.tsx]: seems to be unused

File: examples/LabeledMerge.tsx (hunk 1 above)
Comment (jh3y) [reply examples/LabeledMerge.tsx]: Thanks 👍

File: src/types.ts [hunk 3]
```diff
@@ -48,6 +48,11 @@ export interface ITrackProps {
   onTouchStart: (e: React.TouchEvent) => void;
 }
 
+export interface IRangeRef {
+  getOffsets: () => any;
+  getThumbs: () => any;
```
Comment (tajo) [inline src/types.ts]:
```
getThumbs: () => Element[];
```

?

File: src/utils.ts [hunk 4]
```diff
@@ -142,3 +143,106 @@ export function voidFn() {}
...
+/**
+ * A custom React Hook for calculating whether a thumb overlaps
+ * another and whether labels could/should merge.
+ * @param rangeRef - React ref value of Range component
+ * @param values - current Range values Array
+ * @param index - thumb index
+ * @param separator - string to separate thumb values
+ */
+export const useThumbOverlap = (
+  rangeRef: IRangeRef | undefined,
```
Comment (tajo) [inline src/utils.ts]:
```suggestion
  rangeRef?: IRangeRef,
```

Comment (tajo) [reply src/utils.ts]: Also, if this can be undefined, should it be the last param?

File: src/types.ts [hunk 5]
```diff
@@ -48,6 +48,11 @@ export interface ITrackProps {
   onTouchStart: (e: React.TouchEvent) => void;
 }
 
+export interface IRangeRef {
+  getOffsets: () => any;
```
Comment (tajo) [inline src/types.ts]: Can the return be typed?

File: package.json [hunk 6]
```diff
@@ -38,17 +38,18 @@
     "interval"
   ],
   "devDependencies": {
-    "@storybook/addon-options": "4.2.0-alpha.8",
-    "@storybook/addons": "4.2.0-alpha.8",
-    "@storybook/react": "4.2.0-alpha.8",
+    "babel-loader": "^8.0.6",
```
Comment (tajo) [inline package.json]: what's up with `babel`?

File: src/utils.ts [hunk 7]
```diff
@@ -142,3 +143,106 @@ export function voidFn() {}
...
+  values: number[],
+  index: number,
+  separator = '-',
+) => {
+  const [style, setStyle] = useState<{ [key: string]: any }>({});
```
Comment (tajo) [inline src/utils.ts]: `React.CSSProperties`

File: src/utils.ts [hunk 8]
```diff
@@ -142,3 +143,106 @@ export function voidFn() {}
...
+  const [labelValue, setLabelValue] = useState(values[index].toFixed(1));
+
+  useEffect(() => {
+    if (rangeRef) {
+      let overlaps = [];
+      let labelValue = values[index].toFixed(1);
+      const offsets = rangeRef.getOffsets();
+      const THUMB_SIZE = rangeRef.getThumbs()[0].offsetWidth;
+      const newStyle: { [key: string]: any } = {};
```
Comment (tajo) [inline src/utils.ts]: `React.CSSProperties`

File: src/utils.ts (hunk 4 above)
Comment (jh3y) [reply src/utils.ts]: The hook doesn't really work without it but the ref value can be `null` the first time the hook is run. The hook won't do anything without the ref value.

File: src/utils.ts [hunk 9]
```diff
@@ -142,3 +143,106 @@ export function voidFn() {}
...
+) => {
+  const [style, setStyle] = useState<{ [key: string]: any }>({});
+  const [labelValue, setLabelValue] = useState(values[index].toFixed(1));
+
+  useEffect(() => {
+    if (rangeRef) {
+      let overlaps = [];
+      let labelValue = values[index].toFixed(1);
+      const offsets = rangeRef.getOffsets();
+      const THUMB_SIZE = rangeRef.getThumbs()[0].offsetWidth;
```
Comment (tajo) [inline src/utils.ts]: I would check if `rangeRef.getThumbs().length > 0`. We throw a warning already but it would nice to prevent the crash.

File: package.json (hunk 6 above)
Comment (jh3y) [reply package.json]: It's required for `storybook` v5. Without it, `storybook` won't build/load. They added something to their docs about it becoming a dependency.

File: src/utils.ts [hunk 10]
```diff
@@ -142,3 +143,106 @@ export function voidFn() {}
...
+  values: number[],
+  index: number,
+  separator = '-',
+) => {
+  const [style, setStyle] = useState<{ [key: string]: any }>({});
+  const [labelValue, setLabelValue] = useState(values[index].toFixed(1));
+
+  useEffect(() => {
+    if (rangeRef) {
+      let overlaps = [];
```
Comment (tajo) [inline src/utils.ts]: `Number[]` ?

File: src/utils.ts [hunk 11]
```diff
@@ -142,3 +143,106 @@ export function voidFn() {}
...
+        const currentX = offsets[c].x;
+        for (let s = 0; s < values.length; s++) {
+          const siblingX = offsets[s].x;
+          if (
+            c !== s &&
+            ((currentX >= siblingX && currentX <= siblingX + THUMB_SIZE) ||
+              (currentX + THUMB_SIZE >= siblingX &&
+                currentX + THUMB_SIZE <= siblingX + THUMB_SIZE))
+          ) {
+            let overlapIndex;
```
Comment (tajo) [inline src/utils.ts]:
```suggestion
            let overlapIndex?: number;
```

File: examples/LabeledMerge.tsx [hunk 12]
```diff
@@ -0,0 +1,142 @@
...
+        style={{
+          display: 'flex',
+          justifyContent: 'center',
+          flexWrap: 'wrap'
+        }}
+      >
+        <Range
+          allowOverlap
+          values={this.state.values}
+          ref={r => (this.RANGE = r)}
```
Comment (tajo) [inline examples/LabeledMerge.tsx]: `this.rangeRef` might be better naming? Also, `React.createRef()` seems to be the preferred way over callbacks since React `16.3`.

**Comment by tajo** [comment]
**I really like the hook, happy to see that we can abstract it!** 

Some notes:

- Can you type all the `any`s? Especially the variable `overlaps` is critical to understand what's happening in the code.
- It's quite hard to follow the hook's code. Some things that could help to make it more readable:

  - Replace `for` with `forEach` to reduce usage of one-letter indexes
  - type the reduce functions
  - don't use one-letter variables

- Are e2e screenshots updated? What happened?
- Can we add e2e tests for this feature?
- Some parts of the hook could be split into a different function, like the part where you build the `overlaps`. 
- Could you add more inline comments explaining the whole algorithm (especially what output you need). I have only a vague idea what it does. There might be some more efficient solution, it's similar to dynamic programming problems. No need to over-optimize it now but for the future reference it would be nice to document what's going on.
- You'll need to update `/flowtypes/utils.js.flow`. I recommend to `yarn link` this in `baseui` and implement the functionality there to test the flow types.

**Comment by jh3y** [comment]
Hey @tajo 👋 
Updates made:
- [x] Typed any existing `any`s 👍 
- [x] Inline commented the hook code + refactored the overlap building. It's much smaller and uses recursion 💪 
- [x] `e2e` snapshots updated with no problems
- [x] Updated `flow` declaration for use in `baseui`. However, `yarn link` doesn't seem to be working as expected for me.
- [x] Added `e2e` tests that test the different overlaps and which labels should be hidden and their values. However, these tests pass on `test:e2e:dev` but fail on `test:e2e`. Is there something I'm missing? I've watched `Chromium` do it's thing and all looks good but when running `test:e2e` it fails on each snapshot. It's like the viewport dimensions are different between the environments.

File: examples/LabeledMerge.tsx [hunk 13]
```diff
@@ -0,0 +1,139 @@
...
+  index,
+}: {
+  rangeRef: Range | null;
+  values: number[];
+  index: number;
+}) => {
+  const [labelValue, style] = useThumbOverlap(rangeRef, values, index);
+  return (
+    <div
+      style={{
```
Comment (gergelyke) [inline examples/LabeledMerge.tsx]: in puppeteer, we want to be able to target these elements, so let's add an identifier:

```
data-label={index}
```

File: e2e/merged.test.ts [hunk 14]
```diff
@@ -0,0 +1,65 @@
...
+
+test('Overlap thumbs 1 and 2', async () => {
+  await trackMouse(page);
+  await page.mouse.move(165, 80);
+  await page.mouse.down();
+  await page.mouse.move(270, 80);
+  await page.mouse.up();
+  await untrackMouse(page);
+  const output = await page.$('[draggable]:nth-of-type(1) div');
+  const hidden = await page.$('[draggable]:nth-of-type(2) div');
```
Comment (gergelyke) [inline e2e/merged.test.ts]:
```
  await page.waitForSelector('[data-label="1"]', {
    hidden: true
  });
```

File: e2e/merged.test.ts [hunk 15]
```diff
@@ -0,0 +1,65 @@
...
+  await page.mouse.down();
+  await page.mouse.move(270, 80);
+  await page.mouse.up();
+  await page.mouse.move(450, 80);
+  await page.mouse.down();
+  await page.mouse.move(330, 80);
+  await page.mouse.up();
+  await untrackMouse(page);
+  const output = await page.$('[draggable]:nth-of-type(1) div');
+  const hidden = await page.$('[draggable]:nth-of-type(2) div');
```
Comment (gergelyke) [inline e2e/merged.test.ts]:
```
  await page.waitForSelector('[data-label="1"]', {
    hidden: true
  });

  await page.waitForSelector('[data-label="2"]', {
    hidden: true
  });

```

File: e2e/merged.test.ts [hunk 16]
```diff
@@ -0,0 +1,65 @@
...
+  await page.mouse.down();
+  await page.mouse.move(270, 80);
+  await page.mouse.up();
+  await page.mouse.move(165, 80);
+  await page.mouse.down();
+  await page.mouse.move(330, 80);
+  await page.mouse.up();
+  await untrackMouse(page);
+  const output = await page.$('[draggable]:nth-of-type(3) div');
+  const hidden = await page.$('[draggable]:nth-of-type(1) div');
```
Comment (gergelyke) [inline e2e/merged.test.ts]:
```
  await page.waitForSelector('[data-label="0"]', {
    hidden: true
  });

  await page.waitForSelector('[data-label="1"]', {
    hidden: true
  });
```

**Comment by jh3y** [comment]
Thanks so much for taking a look at this @gergelyke 👍 
Tests updated and passing 🙌

**Comment by gergelyke** [comment]
@tajo @jh3y can we simply remove the visual diffs if they seem to be unreliable? puppeteer tests the functionality anyways
